File: Catalog REST/src/main/java/cz/vhromada/catalog/rest/controller/GenreController.java
package cz.vhromada.catalog.rest.controller;

import cz.vhromada.catalog.facade.GenreFacade;
import cz.vhromada.catalog.facade.to.GenreTO;
import cz.vhromada.validators.exceptions.RecordNotFoundException;
import cz.vhromada.validators.exceptions.ValidationException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.ResponseBody;

/**
 * A class represents controller for genres.
 *
 * @author Vladimir Hromada
 */
@Controller("genreController")
@RequestMapping("/genres")
public class GenreController extends JsonController {

	@Autowired
	@Qualifier("genreFacade")
	private GenreFacade genreFacade;

	/** Creates new data. */
	@RequestMapping(value = "/new", method = RequestMethod.GET)
	@ResponseBody
	public void newData() {
		genreFacade.newData();
	}

	/**
	 * Returns list of genres.
	 *
	 * @return list of genres
	 */
	@RequestMapping(value = { "", "/", "list" }, method = RequestMethod.GET)
	@ResponseBody
	public String getGenres() {
		return serialize(genreFacade.getGenres());
	}

	/**
	 * Returns genre with ID or null if there isn't such genre.
	 *
	 * @param id ID
	 * @return genre with ID or null if there isn't such genre
	 * @throws IllegalArgumentException if ID is null
	 */
	@RequestMapping(value = "/{id}", method = RequestMethod.GET)
	@ResponseBody
	public String getGenre(@PathVariable("id") final Integer id) {
		return serialize(genreFacade.getGenre(id));
	}

	/**
	 * Adds genre. Sets new ID and position.
	 *
	 * @param genre genre
	 * @throws IllegalArgumentException if genre is null
	 * @throws ValidationException      if ID isn't null
	 *                                  or name is null
	 *                                  or name is empty string
	 */
	@RequestMapping(value = "/add", method = RequestMethod.POST)
	@ResponseBody
	public void add(final String genre) {
		genreFacade.add(deserialize(genre, GenreTO.class));
	}

	/**
	 * Adds list of genre names.
	 *
	 * @param genre genre
	 * @throws IllegalArgumentException if list of genre names is null
	 * @throws ValidationException      if list of genre names contains null value
	 */
	@RequestMapping(value = "/addList", method = RequestMethod.POST)
	@ResponseBody
	public void addList(final String genre) {
		genreFacade.add(this.<String>deserializeList(genre));
	}

	/**
	 * Updates genre.
	 *
	 * @param genre new value of genre
	 * @throws IllegalArgumentException if genre is null
	 * @throws ValidationException      if ID is null
	 *                                  or name is null
	 *                                  or name is empty string
	 * @throws RecordNotFoundException  if genre doesn't exist in data storage
	 */
	@RequestMapping(value = "/update", method = RequestMethod.POST)
	@ResponseBody
	public void update(final String genre) {
		genreFacade.update(deserialize(genre, GenreTO.class));
	}

	/**
	 * Removes genre.
	 *
	 * @param genre genre
	 * @throws IllegalArgumentException if genre is null
	 * @throws ValidationException      if ID is null
	 * @throws RecordNotFoundException  if genre doesn't exist in data storage
	 */
	@RequestMapping(value = "/remove", method = RequestMethod.POST)
	@ResponseBody
	public void remove(final String genre) {
		genreFacade.remove(deserialize(genre, GenreTO.class));
	}

	/**
	 * Duplicates genre.
	 *
	 * @param genre genre
	 * @throws IllegalArgumentException if genre is null
	 * @throws ValidationException      if ID is null
	 * @throws RecordNotFoundException  if genre doesn't exist in data storage
	 */
	@RequestMapping(value = "/duplicate", method = RequestMethod.POST)
	@ResponseBody
	public void duplicate(final String genre) {
		genreFacade.duplicate(deserialize(genre, GenreTO.class));
	}

	/**
	 * Returns true if genre exists.
	 *
	 * @param genre genre
	 * @return true if genre exists
	 * @throws IllegalArgumentException if genre is null
	 * @throws ValidationException      if ID is null
	 */
	@RequestMapping(value = "/exists", method = RequestMethod.GET)
	@ResponseBody
	public String exists(final String genre) {
		return serialize(genreFacade.exists(deserialize(genre, GenreTO.class)));
	}

}

File: catalog-rest/src/main/kotlin/com/github/vhromada/catalog/rest/controller/GenreController.kt
package com.github.vhromada.catalog.rest.controller

import com.github.vhromada.catalog.entity.Genre
import com.github.vhromada.catalog.facade.GenreFacade
import com.github.vhromada.common.web.controller.AbstractController
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestController

/**
 * A class represents controller for genres.
 *
 * @author Vladimir Hromada
 */
@RestController("genreController")
@RequestMapping("/catalog/genres")
class GenreController(
    private val genreFacade: GenreFacade
) : AbstractController() {

    /**
     * Returns list of genres.
     *
     * @return list of genres
     */
    @GetMapping
    fun getGenres(): List<Genre> {
        return processResult(result = genreFacade.getAll())!!
    }

    /**
     * Creates new data.
     */
    @PostMapping("/new")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun newData() {
        processResult(result = genreFacade.newData())
    }

    /**
     * Returns genre with ID.
     *
     * @param id ID
     * @return genre with ID
     */
    @GetMapping("/{id}")
    fun getGenre(@PathVariable("id") id: Int): Genre {
        return processResult(result = genreFacade.get(id = id))!!
    }

    /**
     * Adds genre. Sets new ID and position.
     * <br></br>
     * Validation errors:
     *
     *  * ID isn't null
     *  * Position isn't null
     *  * Name is null
     *  * Name is empty string
     *
     * @param genre genre
     */
    @PutMapping("/add")
    @ResponseStatus(HttpStatus.CREATED)
    fun add(@RequestBody genre: Genre) {
        processResult(result = genreFacade.add(data = genre))
    }

    /**
     * Updates genre.
     * <br></br>
     * Validation errors:
     *
     *  * ID is null
     *  * Position is null
     *  * Name is null
     *  * Name is empty string
     *  * Genre doesn't exist in data storage
     *
     * @param genre new value of genre
     */
    @PostMapping("/update")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun update(@RequestBody genre: Genre) {
        processResult(result = genreFacade.update(data = genre))
    }

    /**
     * Removes genre.
     * <br></br>
     * Validation errors:
     *
     *  * Genre doesn't exist in data storage
     *
     * @param id ID
     */
    @DeleteMapping("/remove/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun remove(@PathVariable("id") id: Int) {
        processResult(result = genreFacade.remove(id = id))
    }

    /**
     * Duplicates genre.
     * <br></br>
     * Validation errors:
     *
     *  * Genre doesn't exist in data storage
     *
     * @param id ID
     */
    @PostMapping("/duplicate/{id}")
    @ResponseStatus(HttpStatus.CREATED)
    fun duplicate(@PathVariable("id") id: Int) {
        processResult(result = genreFacade.duplicate(id = id))
    }

    /**
     * Moves genre in list one position up.
     * <br></br>
     * Validation errors:
     *
     *  * Genre can't be moved up
     *  * Genre doesn't exist in data storage
     *
     * @param id ID
     */
    @PostMapping("/moveUp/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun moveUp(@PathVariable("id") id: Int) {
        processResult(result = genreFacade.moveUp(id = id))
    }

    /**
     * Moves genre in list one position down.
     * <br></br>
     * Validation errors:
     *
     *  * Genre can't be moved down
     *  * Genre doesn't exist in data storage
     *
     * @param id ID
     */
    @PostMapping("/moveDown/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun moveDown(@PathVariable("id") id: Int) {
        processResult(result = genreFacade.moveDown(id = id))
    }

    /**
     * Updates positions.
     */
    @PostMapping("/updatePositions")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun updatePositions() {
        processResult(result = genreFacade.updatePositions())
    }

}
